// terminal-query/index.js
import {Buffer} from 'node:buffer';
import fs from 'node:fs';
import process from 'node:process';
import {StringDecoder} from 'node:string_decoder';
import tty from 'node:tty';

const matchesResponseTerminator = (buffer, responseTerminator) => {
	if (!responseTerminator) {
		return false;
	}

	if (typeof responseTerminator === 'string') {
		return responseTerminator.length > 0 && buffer.includes(responseTerminator);
	}

	// Clone to avoid mutating the user's regex lastIndex.
	// Sticky regexes require lastIndex to be copied since they match at that exact position.
	// Global regexes intentionally start from 0 to search the entire buffer.
	const matcher = new RegExp(responseTerminator);
	if (responseTerminator.sticky) {
		matcher.lastIndex = responseTerminator.lastIndex;
	}

	return matcher.test(buffer);
};

const toHexString = value => Buffer.from(value).toString('hex');

const normalizeCapabilityNames = capabilityNames =>
	Array.isArray(capabilityNames) ? capabilityNames : [capabilityNames];

const normalizeHelperOptions = (options, responseTerminator) => {
	if (!options) {
		return {responseTerminator};
	}

	const {responseTerminator: _responseTerminator, ...rest} = options;
	return {
		...rest,
		responseTerminator,
	};
};

// eslint-disable-next-line no-control-regex
const stringTerminatorPattern = /\u001B\\|\u009C/;
// eslint-disable-next-line no-control-regex
const oscTerminatorPattern = /\u0007|\u001B\\|\u009C/;

const createXTGETTCAPQuery = capabilityNames => {
	const normalizedCapabilityNames = normalizeCapabilityNames(capabilityNames).filter(Boolean);

	if (normalizedCapabilityNames.length === 0) {
		return;
	}

	const encodedNames = normalizedCapabilityNames.map(value => toHexString(value)).join(';');
	return `\u001BP+q${encodedNames}\u001B\\`;
};

const extractXTVERSIONPayload = response => {
	if (!response) {
		return;
	}

	// The payload capture stops at the String Terminator (ESC \ or 0x9C).
	// eslint-disable-next-line no-control-regex
	const match = response.match(/\u001BP>\|([^\u001B\u009C]*)(?:\u001B\\|\u009C)/);
	return match?.[1];
};

/**
Send an escape sequence to the terminal and return the response.

@param {string} query - The escape sequence to send.
@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {string | RegExp} [options.responseTerminator='\u001B\\'] - When the buffer matches the terminator, the query finishes early.
@param {NodeJS.ReadStream} [options.stdin=process.stdin] - Input stream to read from.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {Promise<string | undefined>} The response, or `undefined` if not in a TTY or no response is received.
*/
export async function queryTerminal(
	query,
	{
		timeoutMilliseconds = 100,
		responseTerminator = '\u001B\\',
		stdin = process.stdin,
		stdout = process.stdout,
	} = {},
) {
	if (!stdin.isTTY || !stdout.isTTY) {
		return;
	}

	const wasFlowing = stdin.readableFlowing;
	const shouldResume = wasFlowing === false;
	const shouldRestoreNullFlowing = wasFlowing === null;
	const existingListenerCount = stdin.listenerCount('data') + stdin.listenerCount('readable');
	const shouldUnref = typeof stdin.unref === 'function' && wasFlowing !== true && existingListenerCount === 0;
	const previousRawMode = stdin.isRaw ?? false;
	const didEnableRawMode = !previousRawMode;
	const decoder = new StringDecoder('utf8');

	let buffer = '';
	let timeout;
	let isFinished = false;

	return new Promise((resolve, reject) => {
		const cleanup = () => {
			if (isFinished) {
				return false;
			}

			isFinished = true;

			clearTimeout(timeout);
			stdin.off('data', onData);
			stdin.off('error', onError);

			buffer += decoder.end();

			if (shouldRestoreNullFlowing) {
				try {
					stdin.readableFlowing = null;
				} catch {
					// Ignore if the stream does not allow resetting readableFlowing.
				}
			} else if (shouldResume) {
				stdin.pause();
			}

			if (didEnableRawMode) {
				try {
					stdin.setRawMode(false);
				} catch {}
			}

			if (shouldUnref) {
				try {
					stdin.unref();
				} catch {}
			}

			return true;
		};

		const finish = () => {
			if (!cleanup()) {
				return;
			}

			resolve(buffer || undefined);
		};

		const fail = error => {
			if (cleanup()) {
				reject(error);
			}
		};

		const onData = data => {
			buffer += typeof data === 'string' ? data : decoder.write(data);

			if (matchesResponseTerminator(buffer, responseTerminator)) {
				finish();
			}
		};

		const onError = error => {
			fail(error);
		};

		try {
			if (didEnableRawMode) {
				stdin.setRawMode(true);
			}

			if (shouldResume) {
				stdin.resume();
			}

			stdin.on('data', onData);
			stdin.on('error', onError);
			stdout.write(query);

			timeout = setTimeout(() => {
				finish();
			}, timeoutMilliseconds);
		} catch (error) {
			fail(error);
		}
	});
}

/**
Query terminal name and version (XTVERSION).

@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.ReadStream} [options.stdin=process.stdin] - Input stream to read from.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {Promise<string | undefined>} The parsed version string, or `undefined` if not in a TTY or parsing fails.
*/
export async function queryXTVERSION(options) {
	return extractXTVERSIONPayload(await queryTerminal('\u001B[>q', normalizeHelperOptions(options, stringTerminatorPattern)));
}

/**
Query terminfo capabilities (XTGETTCAP).

@param {string | string[]} capabilityNames - Terminfo capability names (for example: "indn").
@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.ReadStream} [options.stdin=process.stdin] - Input stream to read from.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {Promise<string | undefined>} The response, or `undefined` if not in a TTY or no response is received.
*/
export function queryXTGETTCAP(capabilityNames, options) {
	const query = createXTGETTCAPQuery(capabilityNames);

	if (!query) {
		throw new TypeError('Expected at least one capability name');
	}

	return queryTerminal(query, normalizeHelperOptions(options, stringTerminatorPattern));
}

/**
Query primary device attributes (DA1).

@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.ReadStream} [options.stdin=process.stdin] - Input stream to read from.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {Promise<string | undefined>} The response, or `undefined` if not in a TTY or no response is received.
*/
export function queryDeviceAttributes(options) {
	return queryTerminal('\u001B[c', normalizeHelperOptions(options, /c$/));
}

/**
Query iTerm feature reporting (OSC 1337;Capabilities).

@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.ReadStream} [options.stdin=process.stdin] - Input stream to read from.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {Promise<string | undefined>} The response, or `undefined` if not in a TTY or no response is received.
*/
export function queryITermFeatures(options) {
	return queryTerminal('\u001B]1337;Capabilities\u001B\\', normalizeHelperOptions(options, oscTerminatorPattern));
}

/**
Send an escape sequence to the terminal and return the response synchronously.

@param {string} query - The escape sequence to send.
@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {string | RegExp} [options.responseTerminator='\u001B\\'] - When the buffer matches the terminator, the query finishes early.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {string | undefined} The response, or `undefined` if not in a TTY or no response is received.
*/
export function queryTerminalSync(
	query,
	{
		timeoutMilliseconds = 100,
		responseTerminator = '\u001B\\',
		stdout = process.stdout,
	} = {},
) {
	if (!stdout.isTTY) {
		return;
	}

	let fd;
	try {
		// eslint-disable-next-line no-bitwise
		fd = fs.openSync('/dev/tty', fs.constants.O_RDONLY | fs.constants.O_NONBLOCK);
	} catch {
		return;
	}

	let input;
	try {
		input = new tty.ReadStream(fd);
	} catch {
		fs.closeSync(fd);
		return;
	}

	const previousRawMode = process.stdin.isRaw ?? false;
	const didEnableRawMode = !previousRawMode;
	let buffer = '';
	const readBuffer = Buffer.alloc(256);
	const decoder = new StringDecoder('utf8');
	const waitBuffer = new Int32Array(new SharedArrayBuffer(4));
	const deadline = Date.now() + timeoutMilliseconds;

	try {
		if (didEnableRawMode) {
			input.setRawMode(true);
		}

		stdout.write(query);

		while (Date.now() < deadline) {
			let bytesRead = 0;
			try {
				bytesRead = fs.readSync(fd, readBuffer, 0, readBuffer.length, null);
			} catch (error) {
				if (error.code === 'EAGAIN' || error.code === 'EWOULDBLOCK') {
					// No data available yet, brief sleep to avoid CPU spinning
					Atomics.wait(waitBuffer, 0, 0, 1);
					continue;
				}

				throw error;
			}

			if (bytesRead > 0) {
				buffer += decoder.write(readBuffer.subarray(0, bytesRead));

				if (matchesResponseTerminator(buffer, responseTerminator)) {
					break;
				}
			}
		}

		// Flush any remaining bytes in the decoder
		buffer += decoder.end();
	} finally {
		if (didEnableRawMode) {
			try {
				input.setRawMode(false);
			} catch {}
		}

		input.destroy();
		try {
			fs.closeSync(fd);
		} catch {
			// Ignore if the stream already closed the file descriptor.
		}
	}

	return buffer || undefined;
}

/**
Query terminal name and version (XTVERSION) synchronously.

@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {string | undefined} The parsed version string, or `undefined` if not in a TTY or parsing fails.
*/
export function queryXTVERSIONSync(options) {
	return extractXTVERSIONPayload(queryTerminalSync('\u001B[>q', normalizeHelperOptions(options, stringTerminatorPattern)));
}

/**
Query terminfo capabilities (XTGETTCAP) synchronously.

@param {string | string[]} capabilityNames - Terminfo capability names (for example: "indn").
@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {string | undefined} The response, or `undefined` if not in a TTY or no response is received.
*/
export function queryXTGETTCAPSync(capabilityNames, options) {
	const query = createXTGETTCAPQuery(capabilityNames);

	if (!query) {
		throw new TypeError('Expected at least one capability name');
	}

	return queryTerminalSync(query, normalizeHelperOptions(options, stringTerminatorPattern));
}

/**
Query primary device attributes (DA1) synchronously.

@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {string | undefined} The response, or `undefined` if not in a TTY or no response is received.
*/
export function queryDeviceAttributesSync(options) {
	return queryTerminalSync('\u001B[c', normalizeHelperOptions(options, /c$/));
}

/**
Query iTerm feature reporting (OSC 1337;Capabilities) synchronously.

@param {object} options
@param {number} [options.timeoutMilliseconds=100] - Time to wait for response.
@param {NodeJS.WriteStream} [options.stdout=process.stdout] - Output stream to write to.
@returns {string | undefined} The response, or `undefined` if not in a TTY or no response is received.
*/
export function queryITermFeaturesSync(options) {
	return queryTerminalSync('\u001B]1337;Capabilities\u001B\\', normalizeHelperOptions(options, oscTerminatorPattern));
}
