// sixel/lib/bundle_decode.js
"use strict";
/**
 * Copyright (c) 2021 Joerg Breitbart.
 * @license MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PALETTE_VT340_GREY = exports.PALETTE_VT340_COLOR = exports.PALETTE_ANSI_256 = exports.fromRGBA8888 = exports.toRGBA8888 = exports.decodeAsync = exports.decode = exports.DecoderAsync = exports.Decoder = void 0;
var Decoder_1 = require("./Decoder");
Object.defineProperty(exports, "Decoder", { enumerable: true, get: function () { return Decoder_1.Decoder; } });
Object.defineProperty(exports, "DecoderAsync", { enumerable: true, get: function () { return Decoder_1.DecoderAsync; } });
Object.defineProperty(exports, "decode", { enumerable: true, get: function () { return Decoder_1.decode; } });
Object.defineProperty(exports, "decodeAsync", { enumerable: true, get: function () { return Decoder_1.decodeAsync; } });
var Colors_1 = require("./Colors");
Object.defineProperty(exports, "toRGBA8888", { enumerable: true, get: function () { return Colors_1.toRGBA8888; } });
Object.defineProperty(exports, "fromRGBA8888", { enumerable: true, get: function () { return Colors_1.fromRGBA8888; } });
Object.defineProperty(exports, "PALETTE_ANSI_256", { enumerable: true, get: function () { return Colors_1.PALETTE_ANSI_256; } });
Object.defineProperty(exports, "PALETTE_VT340_COLOR", { enumerable: true, get: function () { return Colors_1.PALETTE_VT340_COLOR; } });
Object.defineProperty(exports, "PALETTE_VT340_GREY", { enumerable: true, get: function () { return Colors_1.PALETTE_VT340_GREY; } });
